crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -1,88 +1,116 @@
+use std::collections::VecDeque;
+use std::ops::{Deref, DerefMut};
+
 use super::Type;
 
-/// Typed arguments for a single call, in source order.
+/// Arguments for a single call, in source order.
 #[derive(Clone, Debug, Default)]
-pub(crate) struct CallArguments<'a, 'db>(Vec<Argument<'a, 'db>>);
+pub(crate) struct CallArguments<'a>(VecDeque<Argument<'a>>);
 
-impl<'a, 'db> CallArguments<'a, 'db> {
-    /// Create a [`CallArguments`] with no arguments.
-    pub(crate) fn none() -> Self {
-        Self(Vec::new())
+impl<'a> CallArguments<'a> {
+    /// Push an extra synthetic argument (for a `self` or `cls` parameter) to the front of this
+    /// argument list.
+    pub(crate) fn push_self(&mut self) {
+        self.0.push_front(Argument::Synthetic);
     }
 
-    /// Create a [`CallArguments`] from an iterator over non-variadic positional argument types.
-    pub(crate) fn positional(positional_tys: impl IntoIterator<Item = Type<'db>>) -> Self {
-        positional_tys
-            .into_iter()
-            .map(Argument::Positional)
-            .collect()
+    /// Pop the extra synthetic argument from the front of this argument list.
+    pub(crate) fn pop_self(&mut self) {
+        self.0.pop_front();
```

Comment:
Do we need an assertion that the front attribute is `Synthetic` to avoid cases where someone calls `pop_self` without having called `push_self` before? 

Or should we use an `Option` for the `self` argument (I don't know if that's the approach that you had initially), as it seems that we never need to return a slice (which wouldn't work with `VecDeque` anyway) and we only ever need to mutate the first element.

Reply:
I changed this to a `with_self` method that does the pushing/popping in the right pattern, instead of adding assertions or more complex types to ensure the caller does the right thing.